x-pack/plugins/ingest_pipelines/public/url_generator.test.ts

```diff
@@ -0,0 +1,108 @@
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.LIST,
+      });
+      expect(url).toBe('/app/test_app/');
+    });
+
+    it('should generate correct absolute url for list without pipelineId', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.LIST,
+        absolute: true,
+      });
+      expect(url).toBe('http://localhost/app/test_app/');
+    });
+    it('should generate correct relative url for list with a pipelineId', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.LIST,
+        pipelineId: 'pipeline_name',
+      });
+      expect(url).toBe('/app/test_app/?pipeline=pipeline_name');
+    });
+
+    it('should generate correct absolute url for list with a pipelineId', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.LIST,
+        pipelineId: 'pipeline_name',
+        absolute: true,
+      });
+      expect(url).toBe('http://localhost/app/test_app/?pipeline=pipeline_name');
+    });
+  });
+
+  describe('Pipeline Edit', () => {
+    it('should generate correct relative url for pipeline edit', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.EDIT,
+        pipelineId: 'pipeline_name',
+      });
+      expect(url).toBe('/app/test_app/edit/pipeline_name');
+    });
+
+    it('should generate correct absolute url for pipeline edit', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.EDIT,
+        pipelineId: 'pipeline_name',
+        absolute: true,
+      });
+      expect(url).toBe('http://localhost/app/test_app/edit/pipeline_name');
+    });
+  });
+
+  describe('Pipeline Clone', () => {
+    it('should generate correct relative url for pipeline clone', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.CLONE,
+        pipelineId: 'pipeline_name',
+      });
+      expect(url).toBe('/app/test_app/create/pipeline_name');
+    });
+
+    it('should generate correct absolute url for pipeline clone', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.CLONE,
+        pipelineId: 'pipeline_name',
+        absolute: true,
+      });
+      expect(url).toBe('http://localhost/app/test_app/create/pipeline_name');
+    });
+  });
+
+  describe('Pipeline Create', () => {
+    it('should generate correct relative url for pipeline create', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.CREATE,
+        pipelineId: 'pipeline_name',
+      });
+      expect(url).toBe('/app/test_app/create');
+    });
+
+    it('should generate correct absolute url for pipeline clone', async () => {
+      const url = await urlGenerator.createUrl({
+        page: INGEST_PIPELINES_PAGES.CREATE,
```

Comment:
Do consumers of the generator have access to `INGEST_PIPELINES_PAGES`? Is it defined on the public interface somewhere? I wonder why it's not exposed as part of the URL generator's interface, like this:

```js
urlGenerator.createUrl({
  page: urlGenerator.CREATE,
  pipelineId: 'pipeline_name',
});
```

I guess I'm just confused about how consumers in another plugin will know which strings to pass in for the `page` property.

Reply:
We've faced the same issue, it's not possible ATM to extend `urlGenerator` class, check https://github.com/elastic/kibana/blob/543481ba539b5003bbb8f7b2ed987a7c747c964d/src/plugins/share/public/url_generators/url_generator_internal.ts#L70

Reply:
I am currently exporting generator id, generator class, generator state and pages enum for external use in [`public.ts`](https://github.com/yuliacech/kibana/blob/ingest_url_gen/x-pack/plugins/ingest_pipelines/public/url_generator.ts)

```
export {
  INGEST_PIPELINES_APP_ULR_GENERATOR,
  IngestPipelinesUrlGenerator,
  IngestPipelinesUrlGeneratorState,
  INGEST_PIPELINES_PAGES,
} from './url_generator';
```

I think, it should be enough for ML plugin to use this url generator, WDYT @darnautov ?

Reply:
I don't think we need any of these TBH, because you register the URL generator during your Inget Pipeline plugin setup and we simply get it using the `share`. Typescript will validate the provided state as well.